service/raft/raft_group0.cc

```diff
@@ -671,13 +672,25 @@ future<> raft_group0::setup_group0(
         db::system_keyspace& sys_ks, const std::unordered_set<gms::inet_address>& initial_contact_nodes, shared_ptr<group0_handshaker> handshaker,
         std::optional<replace_info> replace_info, service::storage_service& ss, cql3::query_processor& qp, service::migration_manager& mm, bool topology_change_enabled) {
     if (!co_await use_raft()) {
+        // The node is in the RECOVERY mode.
         co_return;
     }
 
-    if (sys_ks.bootstrap_complete()) {
-        // If the node is bootstrapped the group0 server should be setup already
+    if (_sys_ks.bootstrap_complete() && !ss.raft_topology_change_enabled()) {
+        // The node is restarting in the gossip-based topology. There are two possible cases:
+        // - the node is performing a normal restart and group 0 has already been set up,
+        // - the node has just left the RECOVERY mode and the internal upgrade-to-raft procedure will set up group 0
+        // in finish_setup_after_join.
+        co_return;
+    }
+
+    if (joined_group0()) {
+        // Group 0 is already set up, there is nothing to do.
         co_return;
     }
+    // Reaching this point is possible only in two cases:
+    // - the node is bootstrapping,
+    // - the node is restarting in the Raft Recovery procedure and has not joined the new group 0 yet.
```

Comment:
above you mention recovery mode, here you mention recovery procedure, but these refer to different things, it could be confusing

let's disambiguate, and whenever you refer to recovery mode or recovery procedure in the comments, say "old" or "new"

Reply:
Lets find better way to disambiguate :) May be "gossiper base recovery" and "raft topology recovery" or something.

Reply:
I chose "the gossip-based recovery procedure" and "the Raft-based recovery procedure". I updated the names everywhere (cover letter, commit messages, comments, logs).